compliance/collection/intervals/intervals.go

```diff
@@ -0,0 +1,71 @@
+package intervals
+
+import (
+	"math"
+	"math/rand"
+	"time"
+
+	"github.com/stackrox/rox/pkg/env"
+	"github.com/stackrox/rox/pkg/logging"
+)
+
+var (
+	log = logging.LoggerForModule()
+
+	// Used for testing purposes, to mock random values.
+	randFloat64 = rand.Float64
```

Comment:
@vikin91 

> I do not like that this is provided as a dependency implicitly and can be changed by accident

Can you elaborate on what implicit dependency means and in what scenarios this could be changed by accident? I want to evaluate the current solution's risks and problems to ensure your suggestion addresses them.

Reply:
> The implicit behavior here is a result of the NodeScanIntervals depending on existence of a global (here: pkg-level) function outside of the struct. It feels that the NodeScanIntervals depends on the function randFloat64 but it is assumed that this dependency exists outside of this struct.

Correct, except I wouldn't call it implicit because the dependency is explicit since this is a package-level function.

> By suggesting to make this dependency explicit - i.e., passed as a parameter in constructor - we would cause that every programmer that uses NodeScanIntervals would need to think about that function when creating an object of type NodeScanIntervals.

Passing a parameter to the constructor makes the dependency injectable. We would cause every programmer to be able to select which random function the `NodeScanIntervals` would use, and I don't want that. I want them to be safe. I want that to be an implementation detail. I would use an injectable dependency if that was part of the contract of the type. Otherwise, every toggle, every parameter, and every observed behavior can (and will, given an undefined but finite amount of time) be used as a dependency in client code.

> Currently the users of NewNodeScanIntervalFromEnv() are not even aware that having a function called randFloat64 may sneakily change the behavior of NodeScanIntervals and debugging this in the future may be a bit tricky.

Users of the package cannot overwrite the package-level variable.

For the reasons above, I am in favor of keeping the reference to a package level.

Reply:
I think we went a bit too far thinking that every programmer would be able to select the random function. The typical users would use `func NewNodeScanIntervalFromEnv() NodeScanIntervals` that provides a default value that is the same as you provide in the package-level variable.

I do not intend to change this, I want that the control over that function stays in the package. However, I think that using it like this

```go
// in intervals.go
// for public use
func NewNodeScanIntervalFromEnv() NodeScanIntervals { 
	return newNodeScanInterval(/* reasonable defaults or params here */, rand.Float64) 
}

// for private use
func newNodeScanInterval(base time.Duration, deviation int64, initialMax time.Duration, fn func() float64) *NodeScanIntervals { 
	return &NodeScanIntervals{
		base: base,
		deviation: deviation,
		initialMax: initialMax,
		randFloat64 = fn 
	} 
}

// in interval_test.go
func (s *IntervalsTestSuite) Test_Next() {
	tests := []struct {
		name       string
		deviation  int64
 		base       time.Duration
		rand       float64
		initialMax time.Duration
		want       time.Duration
	}{
		// shortened for brevity
		{
			name: "deviation set to 0",
			deviation: 0,
			base:  time.Hour,
			rand: 0,
			initialMax: // whatever the value should be
			want: time.Hour,
		},
	}

	for _, tt := range tests {
		s.Run(tt.name, func() {
			got := newNodeScanInterval(tt.base, tt.deviation, tt.initialMax, tt.rand).Next()
			s.Assert().Equal(tt.want, got)
		})
	}
}

``` 

instead of this

```go
func (s *IntervalsTestSuite) Test_Next() {
	defaultIntervals := NodeScanIntervals{
		base:      time.Hour * 10,
		deviation: 0.1,
	}
	tests := []struct {
		name      string
		intervals NodeScanIntervals
		rand      float64
		want      time.Duration
	}{
		// shortened for brevity
		{
			name: "deviation set to 0",
			intervals: NodeScanIntervals{
				deviation: 0,
				base:      time.Hour,
			},
			rand: 0,
			want: time.Hour,
		},
	}

	for _, tt := range tests {
		s.Run(tt.name, func() {
			randFloat64 = func() float64 { return tt.rand }
			got := tt.intervals.Next()
			s.Assert().Equal(tt.want, got)
		})
	}
}
```

would look much better and would allow us to be aware that such variable is actually used.

I will not fight for that variant, but I see some benefits over the one that is in the code. Consider this as strong opinion weakly held.

Reply:
> instead of the current code would look much better and allow us to be aware that such variable as randFloat64 is actually used.

I see. It's a matter of maintainability, and then style, which approach is "much better". I would like to hear why you think it is better, but let me share my reasons:

In the current code, we want to replace all the uses of `rand.Float64()` when testing. Using it as an attribute of the struct would force me to pass the struct to all functions in the package. It is more idiomatic to do that when the function uses the relevant data in the struct as part of its purpose

You can think that the package-level variable `randFloat64` is a property of the whole package. Not only the object type defined by struct. As I mentioned, package users can't overwrite `randFloat64`, so encapsulation is checked.

> I will not fight for that variant, but I see some benefits over the one that is in the code. Consider this as strong opinion weakly held.

OK. Thanks for the discussion.